[LangRef] Specify icmp on pointers to only compare address

nikic opened the pull request:
This changes LangRef to specify that pointer icmp only compares the address bits of the pointers. That is, `icmp pred %a, %b` is equivalent to `icmp pred ptrtoaddr(%a), ptrtoaddr(%b)`.

Similarly, it specifies that the `nonnull` attribute requires that the address bits are non-zero.

There are a couple of motivations for this:

* For inequality comparisons, this is really the only sensible semantics. Relational comparison of address and metadata bits as a single integer is generally meaningless (unless the metadata bits are equal).
* This matches (as far as I understand) the behavior of existing CHERI implementations.
* LLVM can only reason about the address bits. These semantics allow pointers with non-address bits to receive essentially the same comparison optimization support as ordinary pointers.

In terms of implementation, this PR adjusts:
 * The AMDGPULowerBufferFatPointers pass.
 * An InstCombine fold that may replace pointers with different non-address bits.
 * The fold that replaces pointers based on dominating pointer equality.

It does not adjust:
 * ISel, because we don't have in-tree targets where we can show a difference.
 * Various icmp+ptrtoint transforms, because we'll have to change this code for ptrtoaddr optimization support anyway, and these changes are tightly related.

Related discussion starting from: https://discourse.llvm.org/t/clarifiying-the-semantics-of-ptrtoint/83987/60?u=nikic

nikic committed:
[LangRef] Specify icmp on pointers to only compare address

This changes LangRef to specify that pointer icmp only compares
the address bits of the pointers. That is, `icmp pred %a, %b`
is equivalent to `icmp pred ptrtoaddr(%a), ptrtoaddr(%b)`.

Similarly, it specifies that the `nonnull` attribute requires that
the address bits are non-zero.

There are a couple of motivations for this:

* For inequality comparisons, this is really the only sensible
  semantics. Relational comparison of address and metadata bits
  as a single integer is generally meaningless (unless the
  metadata bits are equal).
* This matches (as far as I understand) the behavior of existing
  CHERI implementations.
* LLVM can only reason about the address bits. These semantics
  allow pointers with non-address bits to receive essentially the
  same comparison optimization support as ordinary pointers.

In terms of implementation, this adjusts only the
AMDGPULowerBufferFatPointers pass. We don't have in-tree targets
to test the difference in ISel yet. There are a number of
icmp+ptrtoint transforms that will have to be adjusted, but I think
this is better done as part of icmp+ptrtoaddr optimization bringup,
as the changes are tightly related.

Related discussion starting from:
https://discourse.llvm.org/t/clarifiying-the-semantics-of-ptrtoint/83987/60?u=nikic

dtcxzyw reviewed (approved):
LangRef changes look reasonable. I'd leave the AMDGPU part to be checked by its maintainers.

It reminds me that we still use pointer size in computeKnownBits. It should be changed to use the address size.

arichardson reviewed (approved):
The wording looks good to me and matches my understanding of the discussion result. 

How to represent a full comparison including external state in IR can be decided later.

nikic committed:
Adjust icmp handling in pointer replacement fold

nikic commented:
ping @krzysz00 to sign off on the AMDGPU buffer fat pointer change this implies.

nikic committed:
Adjust pointer replacement fold for new semantics

The replacement in ptrtoint is invalid in any case, but we currently
still allow it. However, for the case where the pointer and address
size do not match, we should always reject this, as the non-address
bits may differ.

nikic commented:
@davidchisnall I've now included in this PR the necessary adjustments to the pointer equality replacement fold, to clarify what the impact on "pointer substitutability" with this change is. The only impact is that for pointers where we cannot guarantee the same underlying object, and where the pointer has non-address bits, we can now no longer replace uses inside ptrtoint (but can still replace uses in ptrtoaddr). As CHERI shouldn't be using ptrtoint anyway, that should have no practical impact. And on top of that, the ptrtoint replacement is incorrect anyway and will likely be dropped once we start using ptrtoaddr in key places.

> This is currently the behaviour of CHERI C/C++. As we've had more people trying to write these dialects, we've found that this is the number one source of friction. In CHERIoT, our C++ wrapper type around a CHERI capability uses exact (all bits including tag) and we have found from user feedback that this is a more understandable behaviour.

I've been thinking about this some more, and I think the best way to handle these semantics would be to use an intrinsic for equality comparisons that include the non-address bits and tag, and then try to fold it to an address-only icmp in cases where you can guarantee that the non-address bits / tag are the same. The address-only icmp would then optimize as usual. I don't think there is anything useful in terms of generic optimizations that LLVM could do for an icmp that includes not just the non-address bits but also the hidden state.

nikic commented:
As there hasn't been further feedback here, I plan to merge this soon.

nikic committed:
Merge branch 'main' into langref-icmp-ptr
